static/js/typing.js

```diff
@@ -50,15 +76,27 @@ function get_current_time() {
     return Date.now();
 }
 
-function notify_server_start(user_ids_array) {
-    send_typing_notification_ajax(user_ids_array, "start");
+function notify_server_start(to) {
+    send_typing_notification_based_on_msg_type(to, "start");
 }
 
-function notify_server_stop(user_ids_array) {
-    send_typing_notification_ajax(user_ids_array, "stop");
+function notify_server_stop(to) {
+    send_typing_notification_based_on_msg_type(to, "stop");
 }
 
-export const get_recipient = get_user_ids_array;
+export function get_recipient() {
+    const message_type = compose_state.get_message_type();
+    if (message_type === "private") {
+        return get_user_ids_array();
+    }
+    if (message_type === "stream") {
+        const stream_name = compose_state.stream_name();
+        const stream_id = stream_data.get_stream_id(stream_name);
+        const topic = compose_state.topic();
+        return {stream_id, topic};
+    }
+    return null;
+}
```

Comment:
The `return null` might be better written as a `throw new Error("Unexpected message type");`

Reply:
Okay, changed. looks better now.

Reply:
Cool; you might want the same in `send_typing_notification_based_on_message_type`?

Reply:
Just realised that we do not need to throw any error in both cases:
* The `return null` in case of `get_recipient` is intended. Getting the message type to be something else than private or stream means that we aren't talking to anybody and aren't interested in sending any updates to the server. That'll be taken care of by `typing_status.update` if we send it `null` through `get_recipient`.
* I implemented `send_typing_notification_based_on_message_type` badly by using `compose_state.get_message_type()` to figure out `message_type`. This would lead to bugs. One such is when code in `typing_status` tells this function to send a stoped typing request to server, this will do nothing if the compose box was closed by user. It should instead use the existence of `stream_id` in `to`. 

Made the required changes in each case.